tests if custom job model methods work

ksiedlarek opened the pull request.

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,28 @@
+		if self.job.published_date:
+			published_date_is_not_none = True
+		self.assertEqual(published_date_is_not_none, True)
+		print "Published date:{0}".format(self.job.published_date)
```
This test can be simplified quite a lot:

```
self.assertFalse(self.job.published_date)
self.job.publish()
self.assertTrue(self.job.published_date, "Job has no published date")
```

So we assert that there is no published date first, then publish, then assert that there now is a published date. (When we get to writing the views tests, we can also check that the view now displays the published test)
You can use the optional last argument in the assert to print a message if the test fails.  We don't really care too much about the actual publish date (except for debugging while writing the test)

rebkwok replied:
I'd suggest that you test the negative case too - create a job that has ready_to_publish False, and test that trying to publish it doesn't succeed.
Also test that what happens if you try to publish a job that's already published (with ready_to_publish True and False)

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,28 @@
+	def setUp(self):
+		self.job = mommy.make(Job, ready_to_publish=True)
+
+	def test_is_published(self):
```
Add a doc string to describe what this test is doing (e.g. "Test that a job marked ready to publish has a published_date after publishing)

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,28 @@
+from jobs.models import Job
+
+
+class JobMethodTests(TestCase):
```
Suggest "JobModelTests" (since later we will have View tests and Form tests etc, which may also have methods)

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,28 @@
+		if self.job.expiration_date:
+			expiration_date_is_not_none = True
+		self.assertEqual(expiration_date_is_not_none, True)
+		print "Expiration date:{0}".format(self.job.expiration_date)
```
Could be simplified similarly:

```
self.job.publish()
self.assertFalse(self.job.expiration_date)
self.job.set_expiration_date()
self.assertTrue(self.job.expiration_date, "Job has no expiration date")
```

You could also check that the expiration date is what you expect (find self.job.expiration_date against timezone.now().  This gives you a timedelta object that will be in days, mins and secs - do .days on the timedelta object to get the number of days - it should be 59

rebkwok replied:
Also check other edge cases that you can think of - what happens if you try to set an expiration date when there's one already set, what if you try to set an expiration date on a job that's not ready to publish?  You can also write tests for the cases that we might need in the future - e.g. can you publish without having reviewed?  At the moment we can - but we might want to change that behaviour later.

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,28 @@
+		self.assertEqual(published_date_is_not_none, True)
+		print "Published date:{0}".format(self.job.published_date)
+
+	def test_expiration_date(self):
```
add docstring

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,28 @@
 from django.test import TestCase
 
-# Create your tests here.
+from model_mommy import mommy
```
add model_mommy to requirements.txt

rebkwok commented:
Hi Kasia
I've added a bunch of comments on the tests and also some suggestions for others - basically, test every odd case you can think of that might happen if people try to review/publish etc in the order we didn't intend.  Then when the models change or new methods are written, we can run all the tests and check if the new things inadvertently break something else.
(There's no rush on this though - the tests aren't going to hold up anything else)
Becky

ksiedlarek commented:
ok, I've added some tests, and changed a bit those which exist. I also changed set_expiration_date, but I am not sure if this was the correct way of doing things..

rebkwok commented on jobs/models.py:
```diff
@@ -60,9 +60,12 @@ def publish(self):
+                self.expiration_date = self.published_date + timedelta(60)
+                self.save()
+        except TypeError:
+            return "Error"
```
You get a type error if you do set_expiration_date and there's no published_date, right?  This would be better covered in the if statement, since it's a known case

ksiedlarek replied:
yes, I've changed it.

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,82 @@
+class JobModelTests(TestCase):
+    """Tests for custom methods of Job model."""
+
+    """Setting up models with mommy."""
```
move all docstrings to the line below the function definition (just the standard convention)

ksiedlarek replied:
done.

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,82 @@
+    def test_publish_more_than_once_and_change_values(self):
+        self.assertFalse(self.job.published_date)
+        self.job.publish()
+        self.job.ready_to_publish == False
```
to change the value, use
self.ready_to_publish = False
(self.ready_to_publish == False is just a statement)

ksiedlarek replied:
right, done.

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,82 @@
+        self.assertFalse(self.job.published_date)
+        self.job.publish()
+        self.job.ready_to_publish == False
+        # Right now if we change ready_to_publish to False, published_date stays as it was and doesn't become None.
```
is this what we want to happen?  (might need some discussion about how we expect the ready_to_publish and published_date  to work - once a job is published, do we want the published date to stay the same, or do we want to be able to re-publish?  ).

rebkwok replied:
line too long for PEP8 conventions

ksiedlarek replied:
I am not sure if we want this to happen.. if we republish than date changes. If old date stays until republish we know that it was already published and taken down, it may be good knowledge to have, but also it may be not really useful.

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,82 @@
+        self.assertFalse(self.job.published_date)
+        self.job.publish()
+        self.job.publish()
+        self.assertTrue(self.job.published_date, "Job has no published date.")
```
similar comment to one on previous test - at the moment (I think) publishing twice will give you a new publish date each time (you could check this by assigning published_date to a variable each time and using an assert that compares the two).  Do we want to allow this, or do we want the publish() method to do something different if it already has a publish date?  If we care about it, maybe we need a "last_published_date" so we record the initial published date and the date of the last re-publish?

ksiedlarek replied:
if we had last published date maybe we should just keep track of all what is happening to the job? I mean all of the times something was published and unpublished. But maybe this is too much? Will people really change their minds often? ;)

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,82 @@
+        self.job.publish()
+        self.assertTrue(self.job.published_date, "Job has no published date.")
+
+    """Publish twice in a row."""
```
Expand comment a bit to describe what we're testing happens when we publish twice in a row

ksiedlarek replied:
ok.

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,82 @@
+        self.job.publish()
+        self.assertTrue(self.job.published_date, "Job has no published date.")
+
+    """Publish without review."""
```
Make comment a bit more descriptive.  ALso, at the moment we can publish without review - if we don't want this to happen, or if we need to decide if it can happen, add a #TODO comment (in most text editors it'll be highlighted and will remind us something needs to be looked at for this test

ksiedlarek replied:
done.

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,82 @@
+
+
+    """Test if default value is set."""
+    def test_set_expiration_date(self):
```
this one is redundant, since the next test tests both that a default is set, and what it's set to

ksiedlarek replied:
I removed it

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,82 @@
+    """Test setting expiration date if job is unpublished."""
+    def test_expiration_date_on_not_ready(self):
+        value = self.job.set_expiration_date()
+        self.assertEqual(value, "Error")
```
See comments on the models file.  If a job is unpublished, we just shouldn't be able to set an expiration date. We don't want to be returning a known error.

ksiedlarek replied:
ok, is it good now?

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,82 @@
+        self.job.set_expiration_date()
+        self.assertTrue(self.job.expiration_date, "Job has no expiration date.")
+        self.job.set_expiration_date()
+        self.assertTrue(self.job.expiration_date, "Job has no expiration date.")
```
might want to check the date is changing - i.e. we can extend it by reapplying set_exp_date (should just be able to check that the second date is > the first, or that one minus the other is > 0)

ksiedlarek replied:
it is not - I think I have to republish to get new expiration date.

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,82 @@
+        self.job1.publish()
+        self.assertFalse(self.job1.published_date, "Job has published date.")
+
+    """Publish once, than change ready_to_publish field value and publish once more."""
```
line too long for PEP8 conventions

ksiedlarek replied:
ok, corrected

ksiedlarek commented:
If I don't set expiration date to None, than it doesn't work, dates are the same. But I don't have to set published date to None, so I've deleted it.

rebkwok commented:
Is this a bug in the code?  Should be being set (i.e. extended) each time? Don't set the expiration date back to None in the test to make the test work if it's not something that happens in the code (unless we're expecting users to delete the expiration date before they try to set it again)

ksiedlarek commented:
is it better now?

rebkwok commented on jobs/tests.py:
```diff
@@ -1,3 +1,92 @@
+        self.job.set_expiration_date()
+        second_date = self.job.expiration_date
+        self.assertTrue(self.job.expiration_date, "Job has no expiration date.")
+        self.assertTrue(second_date.second - first_date.second > 0)
```
we can just use

```
self.assertTrue(second_date > first_date)
```

ksiedlarek replied:
ok,done.
